### PR TITLE
[MoE LoRA] fix: Optimizer checkpointing shape mismatch

### PR DESCRIPTION
**Pull request by HollowMan6**
# What does this PR do ?

Resolves https://github.com/volcengine/verl/issues/4303.

# Changelog

- The experts.py code in Megatron-LM set replica_id = (PP, ETP, EDP), but somehow we need to change the ETP rank here with other ids to fit.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to https://github.com/volcengine/verl/issues/4303

<sub>✨ Presented to you with <a href="https://macaron.im/mindlab">Mind Lab</a> - A Lab for Experiential Intelligence.</sub>

### CODE REVIEW

**Comment by HollowMan6** [comment]
Hi @yaoyu-33 ! Though I'm not sure if this PR is the root fix, but this makes the error at https://github.com/volcengine/verl/issues/4303 gone now, and I have tried several parallelism configuration and they are all good. Please let me know what you think.

File: src/megatron/bridge/peft/utils.py [hunk 1]
```diff
@@ -632,6 +632,20 @@ def sharded_state_dict(
...
         linear_out_sd = self.linear_out.sharded_state_dict(f"{prefix}linear_out.", sharded_offsets, metadata)
 
+        # The experts.py code in Megatron-LM set replica_id = (PP, ETP, EDP).
+        if self.is_expert:
+            from megatron.core import parallel_state
+
+            ep_rank = parallel_state.get_expert_model_parallel_rank()
+            edp_rank = parallel_state.get_expert_data_parallel_rank()
+            dp_size = parallel_state.get_data_parallel_world_size()
+            rank = (ep_rank + 1) * (edp_rank + 1) - 1 if dp_size == 1 else ep_rank
```
Comment (yaoyu-33) [inline src/megatron/bridge/peft/utils.py:646]: is this get etp rank?  you can use etp_group = `get_expert_tensor_parallel_group` 
rank = get_pg_rank(etp_group).

from megatron.core.utils import (
    get_pg_rank,
    get_pg_size,
)

Comment (HollowMan6) [reply src/megatron/bridge/peft/utils.py:646]: No, I'm aware that I'm using `edp_rank` here, In mcore it uses `(PP, ETP, EDP)` for `replica_id`, which will cause errors as mentioned in https://github.com/volcengine/verl/issues/4303, since adapter weights are not EP sharded and it assumes that it will replicate along DP modulo EP (sharded by EP).  https://github.com/NVIDIA/Megatron-LM/blob/e8b9df156d5d5f496b21aedbf8af14e91ea036ea/megatron/core/transformer/moe/experts.py#L1160-L1167

Comment (yaoyu-33) [reply src/megatron/bridge/peft/utils.py:646]: This part of comments need to be longer to explain a bit what's happening.

For me, I feel it's still not super clear. If EP is the issue, and how we duplicate and save is the issue, why we are modifying the tp dim(2nd dim) not the dp dim(last dim)?
